packages/labs/analyzer/src/lib/javascript/mixins.ts

```diff
@@ -0,0 +1,194 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * @fileoverview
+ *
+ * Utilities for working with mixins
+ */
+
+import ts from 'typescript';
+import {DiagnosticsError} from '../errors.js';
+import {MixinDeclaration, Analyzer} from '../model.js';
+import {getClassDeclaration} from './classes.js';
+
+const nodeHasMixinHint = (node: ts.Node) =>
+  ts.getJSDocTags(node).some((tag) => tag.tagName.text === 'mixin');
+
+const throwIfMixin = (
+  hasMixinHint: boolean,
+  node: ts.Node,
+  message: string
+) => {
+  if (hasMixinHint) {
+    throw new DiagnosticsError(node, message);
+  } else {
+    return undefined;
+  }
+};
+
+/**
+ * If the given variable declaration was a mixin function, returns a
+ * MixinDeclaration, otherwise returns undefined.
+ *
+ * The mixin logic requires a few important syntactic heuristics to be met in
+ * order to be detected as a mixin.
+ *
+ * If the function is unannotated and does not match the above mixin shape, it
+ * will silently just be analyzed as a simple function and not a mixin. However,
+ * the `@mixin` annotation can be added to produce specific diagnostic errors
+ * when a condition for being analyzed as a mixin is not met.
+ */
+export const maybeGetMixinFromVariableDeclaration = (
+  declaration: ts.VariableDeclaration,
+  analyzer: Analyzer
+): MixinDeclaration | undefined => {
+  const hasMixinHint = nodeHasMixinHint(declaration);
+  const initializer = declaration.initializer;
+  if (
+    initializer === undefined ||
+    !(
+      ts.isArrowFunction(initializer) || ts.isFunctionExpression(initializer)
+    ) ||
+    !ts.isIdentifier(declaration.name)
+  ) {
+    return throwIfMixin(
+      hasMixinHint,
+      initializer ?? declaration,
+      `Expected mixin to be defined as a single const assignment to an ` +
+        `arrow function or function expression.`
+    );
+  }
+  return maybeGetMixinFromFunctionLike(
+    initializer,
+    declaration.name,
+    analyzer,
+    hasMixinHint
+  );
+};
+
+/**
+ * If the given variable declaration was a mixin function, returns a
+ * MixinDeclaration, otherwise returns undefined.
+ *
+ * The mixin logic requires a few important syntactic heuristics to be met in
+ * order to be detected as a mixin.
+ *
+ * If the function is unannotated and does not match the above mixin shape, it
+ * will silently just be analyzed as a simple function and not a mixin. However,
+ * the `@mixin` annotation can be added to produce specific diagnostic errors
+ * when a condition for being analyzed as a mixin is not met.
+ */
+export const maybeGetMixinFromFunctionLike = (
+  fn: ts.FunctionLikeDeclaration,
+  name: ts.Identifier,
+  analyzer: Analyzer,
+  hasMixinHint = nodeHasMixinHint(fn)
+) => {
+  if (!fn.parameters || fn.parameters.length < 1) {
+    return throwIfMixin(
+      hasMixinHint,
+      fn,
+      `Expected mixin to have a superClass parameter.`
```

Comment:
`superClass` spelled and cased that way sounds like it's a specific variable/parameter name. Maybe "have a parameter for the superclass"?